### PR TITLE
Implement nn.functional.interpolate based on upsample.

### PR DESCRIPTION
**Pull request by ailzhang**
This PR addresses #5823. 

* fix docstring: upsample doesn't support LongTensor

* Enable float scale up & down sampling for linear/bilinear/trilinear modes. (following @SsnL 's commit)

* Enable float scale up & down sampling for nearest mode. Note that our implementation is slightly different from TF that there's actually no "align_corners" concept in this mode.

* Add a new interpolate function API to replace upsample. Add deprecate warning for upsample.

* Add an area mode which is essentially Adaptive_average_pooling into resize_image.

* Add test cases for interpolate in test_nn.py

* Add a few comments to help understand *linear interpolation code.

* There is only "*cubic" mode missing in resize_images API which is pretty useful in practice. And it's labeled as hackamonth here #1552. I discussed with @SsnL that we probably want to implement all new ops in ATen instead of THNN/THCUNN. Depending on the priority, I could either put it in my queue or leave it for a HAMer. 

* After the change, the files named as *Upsampling*.c works for both up/down sampling. I could rename the files if needed. 

cc: @SsnL @fmassa

### CODE REVIEW

**Comment by vadimkantorov** [comment]
does it enable bilinear interpolation for ByteTensors as well?

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu [hunk 1]
```diff
@@ -5,152 +5,91 @@
...
            THCTensor *gradOutput,
            THCTensor *gradInput,
-           int scale_factor)
+           int nbatch,
+	   int nchannels,
+	   int inputWidth,
+	   int outputWidth)
 {
-
+  gradInput = THCTensor_(newContiguous)(state, gradInput);
```
Comment (SsnL) [inline aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: Generally you can assume gradInput is contiguous.

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c [hunk 2]
```diff
@@ -2,198 +2,153 @@
 #define TH_GENERIC_FILE "generic/SpatialUpSamplingNearest.c"
 #else
 
+#include "linear_upsampling.h"
+#include <stdio.h>
```
Comment (SsnL) [inline aten/src/THNN/generic/SpatialUpSamplingNearest.c]: Was this for debugging? :)

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c [hunk 3]
```diff
@@ -2,198 +2,153 @@
...
+      int nBatch, int nChannels,
+      int inputHeight, int inputWidth,
+      int outputHeight, int outputWidth) {
+  THArgCheck(inputHeight > 0 && inputWidth > 0
+       && outputHeight > 0 && outputWidth > 0, 2,
+       "input and output sizes should be greater than 0,"
+       " but got input (H: %d, W: %d) output (H: %d, W: %d)",
+       inputHeight, inputWidth, outputHeight, outputWidth);
+  if (input != NULL) {
+    THNN_ARGCHECK(input->_dim() == 4, 2, input,
```
Comment (SsnL) [inline aten/src/THNN/generic/SpatialUpSamplingNearest.c]: Are we dropping support for 3d input?

File: aten/src/THNN/generic/linear_upsampling.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#undef MIN
+#define MIN(a,b) ( ((a)<(b)) ? (a) : (b) )
+#undef MAX
+#define MAX(a,b) ( ((a)>(b)) ? (a) : (b) )
+
+
+template<typename T>
+static inline T linear_upsampling_compute_scale(
+                          int inputSize, int outputSize, bool align_corners) {
+  /* We view each pixel as an area, idx + 0.5 as its center index.
```
Comment (SsnL) [inline aten/src/THNN/generic/linear_upsampling.h]: nice!

File: test/test_nn.py [hunk 5]
```diff
@@ -7030,159 +7083,159 @@ def multimarginloss_weights_no_reduce_test():
         input_size=(1, 9, 4, 4),
     ),
     dict(
-        module_name='Upsample',
+        module_name='Resize_images',
```
Comment (SsnL) [inline test/test_nn.py]: I think the name should be `ResizeImages` according to our convention.

**Review by ssnl — COMMENTED** [review]
Thanks for doing this. The kernel code mostly looks fine, but I didn't spend too much time on it. Here are some general comments:
1. The kernel names are wrong. The NN kernels shouldn't have `interp`. All kernels should have correct dimension number.
2. For the special copying case, can we use `THTensor_(copy)` instead of a nested loop?

File: aten/src/THCUNN/SpatialUpSamplingNearest.cu [hunk 6]
```diff
@@ -2,85 +2,101 @@
...
 
 #include "THCHalf.h"
 #include "THCHalfAutoNumerics.cuh"
+#include "THCAtomics.cuh"
 
-/*
- * Description:
- */
+template<typename Dtype, typename Acctype>
+__global__ void nearest_neighbor_interp2_kernel(
```
Comment (SsnL) [inline aten/src/THCUNN/SpatialUpSamplingNearest.cu]: Why do we need new NN kernels? I remember that you briefly talked about it, but I didn't really understand.

Comment (ailzhang) [reply aten/src/THCUNN/SpatialUpSamplingNearest.cu]: Do you mean we could reuse kernels for linear methods? I guess the main motivation for this was "we had a kernel method for each (linear/nearest , 1d/2d/3d) combination before" and then I thought we might have some reasons to keep specific kernels. But honestly the answer is i'm not sure :P

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu (hunk 1 above)
Comment (ailzhang) [reply aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: Actually all these lines making it contiguous was added in my last commit. There was a segfault in every cuda backward functions if without these lines.

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c (hunk 2 above)
Comment (ailzhang) [reply aten/src/THNN/generic/SpatialUpSamplingNearest.c]: haha you right :P I will remove it.

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c (hunk 3 above)
Comment (ailzhang) [reply aten/src/THNN/generic/SpatialUpSamplingNearest.c]: Yes and no ;) My 2cents : I removed it here bc it was actually never triggered due to a dimension guard in python end. https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1789 And it was never in our python end documentation. 
If we want to support it, we could easily expand the 3D to 4D on python end and pass it along, that way we could avoid complicating our C code by checking sizes. 
What do you think?

**Comment by ailzhang** [comment]
@vadimkantorov Actually all methods like pooling/interpolation don't support byte/half/long tensor. I would prefer to keep it in a separate diff, or even do it along with TH->Aten refactor. cc: @soumith @SsnL @fmassa What do you think?

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu (hunk 1 above)
Comment (SsnL) [reply aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: Hmm that is weird. But they were working before your patch, right?

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c (hunk 3 above)
Comment (SsnL) [reply aten/src/THNN/generic/SpatialUpSamplingNearest.c]: I'm fine with removing. :)

File: aten/src/THCUNN/SpatialUpSamplingNearest.cu (hunk 6 above)
Comment (SsnL) [reply aten/src/THCUNN/SpatialUpSamplingNearest.cu]: Ah okay. I was just mostly wondering why we can't patch the old kernels to make them work.

**Comment by SsnL** [comment]
Yeah I think you should keep those out of this diff.

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu (hunk 1 above)
Comment (ailzhang) [reply aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: Aha I think I get why it's different. So here is context: with the old kernels, it works without initializing gradInput as contiguous. But with the current kernels(like *linear), it segfaults when gradInput is not initialized as contiguous(without last commit). 
And here are my 2 cents: with old code, it didn't initialize as contiguous but get a real * pointer and pass along to the kernel code. In the kernel we actually use the memory contiguously. https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/TemporalUpSamplingNearest.cu#L40
With the new code, we are passing along a THCDeviceTensor, so if it's not initialized as contiguous, it might access some wild memory space during the assignment here. https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/TemporalUpSamplingLinear.cu#L72 
Basically we made sure the memory is contiguous in both cases. And I would prefer to keep it as it is since it's consistent with how *linear methods were implemented. 
What do you think?

File: aten/src/THCUNN/SpatialUpSamplingNearest.cu (hunk 6 above)
Comment (ailzhang) [reply aten/src/THCUNN/SpatialUpSamplingNearest.cu]: Yea I'm not sure about this neither. CC: @fmassa and see if he has any thoughts :D

Comment (SsnL) [reply aten/src/THCUNN/SpatialUpSamplingNearest.cu]: I feel it's fine if you want to change. I was just curious.

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu (hunk 1 above)
Comment (SsnL) [reply aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: I'm not sure the exact reason for segfault. But I realized that you can't call `newContiguous` here as you are overwriting `gradInput` so (if this line actually makes a new tensor) the output is never written to the input of this function!

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c [hunk 7]
```diff
@@ -2,198 +2,152 @@
...
+        const real* pos1 = &idata[h1 * inputWidth + w1];
+        real* pos2 = &odata[h2 * outputWidth + w2];
+        for (int c = 0; c < channels; ++c) {
+          pos2[0] = pos1[0];
+          pos1 += inputHeight * inputWidth;
+          pos2 += outputHeight * outputWidth;
         }
       }
     }
+    return;
```
Comment (fmassa) [inline aten/src/THNN/generic/SpatialUpSamplingNearest.c]: I think you forgot to free the input here

File: aten/src/THCUNN/SpatialUpSamplingNearest.cu [hunk 8]
```diff
@@ -2,85 +2,101 @@
...
+#include "THCDeviceUtils.cuh"
 
 #include "THCHalf.h"
 #include "THCHalfAutoNumerics.cuh"
+#include "THCAtomics.cuh"
 
-/*
- * Description:
- */
+template<typename Dtype, typename Acctype>
```
Comment (fmassa) [inline aten/src/THCUNN/SpatialUpSamplingNearest.cu]: maybe add a comment saying that this kernel is based on the TF implementation, with the link to the implementation?

File: aten/src/THNN/generic/SpatialUpSamplingNearest.c [hunk 9]
```diff
@@ -2,198 +2,152 @@
...
+        real* pos1 = &idata[h1 * inputWidth + w1];
+        const real* pos2 = &odata[h2 * outputWidth + w2];
+        for (int c = 0; c < channels; ++c) {
+          pos1[0] = pos2[0];
+          pos1 += inputHeight * inputWidth;
+          pos2 += outputHeight * outputWidth;
         }
       }
     }
+    return;
```
Comment (fmassa) [inline aten/src/THNN/generic/SpatialUpSamplingNearest.c]: Maybe free the `gradOutput` here as well?

File: aten/src/THNN/generic/TemporalUpSamplingNearest.c [hunk 10]
```diff
@@ -2,172 +2,131 @@
...
+      const int w1 = w2;
+      const real* pos1 = &idata[w1];
+      real* pos2 = &odata[w2];
+      for (int c = 0; c < channels; ++c) {
+        pos2[0] = pos1[0];
+        pos1 += inputWidth;
+        pos2 += outputWidth;
+      }
+    }
+    return;
```
Comment (fmassa) [inline aten/src/THNN/generic/TemporalUpSamplingNearest.c]: free here?

File: aten/src/THNN/generic/TemporalUpSamplingNearest.c [hunk 11]
```diff
@@ -2,172 +2,131 @@
...
+      const int w1 = w2;
+      real* pos1 = &data1[w1];
+      const real* pos2 = &data2[w2];
+      for (int c = 0; c < channels; ++c) {
+        pos1[0] += pos2[0];
+        pos1 += inputWidth;
+        pos2 += outputWidth;
       }
     }
+    return;
```
Comment (fmassa) [inline aten/src/THNN/generic/TemporalUpSamplingNearest.c]: and here?

File: aten/src/THNN/generic/VolumetricUpSamplingNearest.c [hunk 12]
```diff
@@ -2,225 +2,175 @@
...
-              idst += i4*os[4];
-              isrc += iin[4]*is[4];
-            }
-
-            pout[idst] = pin[isrc];
           }
         }
       }
     }
+    return;
```
Comment (fmassa) [inline aten/src/THNN/generic/VolumetricUpSamplingNearest.c]: and here?

File: aten/src/THNN/generic/VolumetricUpSamplingNearest.c [hunk 13]
```diff
@@ -2,225 +2,175 @@
...
+          const real* pos2 = &odata[d2 * outputHeight * outputWidth + h2 * outputWidth + w2];
+          for (int c = 0; c < channels; ++c) {
+            pos1[0] += pos2[0];
+            pos1 += inputDepth * inputHeight * inputWidth;
+            pos2 += outputDepth * outputHeight * outputWidth;
           }
         }
       }
     }
+    return;
```
Comment (fmassa) [inline aten/src/THNN/generic/VolumetricUpSamplingNearest.c]: and here?

**Review by fmassa — CHANGES_REQUESTED** [review]
Hi,

I've had a quick look and I think there are a few memory leaks that should be fixed.

Also, given that we support 1d, 2d and 3d data, I wonder if we should name it `resize_images`, or maybe something a bit more generic?

---
### Not for this PR

For the future, we might want to specify which dimensions we want to interpolate, so that this is not specific to a given data format.

For example
```python
a = torch.rand(1, 3, 100, 100)
A = torch.nn.functional.interpolate(a, axis=[-2. -1], mode='nearest')

# NHWC format
a = torch.rand(1, 100, 100, 3)
A = torch.nn.functional.interpolate(a, axis=[1, 2])

# non-batch mode, for single image
a = torch.rand(3, 100, 100)
A = torch.nn.functional.interpolate(a, axis=[-2, -1])
```

**Comment by vadimkantorov** [comment]
Anothe naming suggestion is `resample` https://github.com/pytorch/pytorch/issues/4277 but maybe it’s a little less discoverable than `interpolate`

File: aten/src/THCUNN/SpatialUpSamplingNearest.cu (hunk 8 above)
Comment (ailzhang) [reply aten/src/THCUNN/SpatialUpSamplingNearest.cu]: Our implementation is not exactly the same as TF as they also have align_corners in the implementation, which doesn't make sense. Not sure if we still want to put the link?

File: aten/src/THCUNN/generic/SpatialUpSamplingNearest.cu [hunk 14]
```diff
@@ -5,165 +5,99 @@
...
+  THAssert(inputHeight > 0 && inputWidth > 0 && outputHeight > 0 && outputWidth > 0);
+
+  THCTensor_(resize4d)(state, output,
+                       THCTensor_(size)(state, input, 0),
+                       THCTensor_(size)(state, input, 1),
+		       outputHeight,
+                       outputWidth);
+  THCTensor_(zero)(state, output);
 
   input = THCTensor_(newContiguous)(state, input);
```
Comment (fmassa) [inline aten/src/THCUNN/generic/SpatialUpSamplingNearest.cu]: question: do we still have the restriction that the input should be contiguous?

File: aten/src/THCUNN/generic/SpatialUpSamplingNearest.cu [hunk 15]
```diff
@@ -5,165 +5,99 @@
...
-  int64_t n_xblocks = min(max((int)ceil((float)no_elements / nthreads), 1), 65535);
-  int64_t n_yblocks = (int64_t)ceil((float)no_elements / (float)(n_xblocks * nthreads));
-  if (n_yblocks > 65535) {
-    THError("Input size is too large!  aborting");
-  }
-  dim3 blocks(n_xblocks, n_yblocks);
-  dim3 threads(nthreads);
+  const int num_kernels = outputHeight * outputWidth;
+  const int num_threads = 
+	  THCState_getCurrentDeviceProperties(state)->maxThreadsPerBlock;
```
Comment (fmassa) [inline aten/src/THCUNN/generic/SpatialUpSamplingNearest.cu]: nit: there is a TAB here

File: aten/src/THCUNN/generic/THCUNN.h [hunk 16]
```diff
@@ -1045,16 +1045,21 @@ THC_API void THNN_(SpatialUpSamplingBilinear_updateGradInput)(
 
 THC_API void THNN_(SpatialUpSamplingNearest_updateGradInput)(
                   THCState *state,
-                  THCTensor *input,
                   THCTensor *gradOutput,
                   THCTensor *gradInput,
-                  int scale_factor);
+                  int nbatch,
+		  int nchannels,
```
Comment (fmassa) [inline aten/src/THCUNN/generic/THCUNN.h]: nit: tabs here and in the next few lines in this file

File: aten/src/THCUNN/generic/TemporalUpSamplingLinear.cu [hunk 17]
```diff
@@ -3,6 +3,7 @@
 #else
 
 #include "../linear_upsampling.h"
+#include <stdio.h>
```
Comment (fmassa) [inline aten/src/THCUNN/generic/TemporalUpSamplingLinear.cu]: nit: do we need this include?

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu [hunk 18]
```diff
@@ -5,151 +5,89 @@
...
-  dim3 threads(nthreads);
+  THCTensor_(resize3d)(state, output,
+                       THCTensor_(size)(state, input, 0),
+                       THCTensor_(size)(state, input, 1),
+                       outputWidth);
+  THCTensor_(zero)(state, output);
 
-  // kernel:
-  upscale<<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (input_data, output_data, no_elements, scale_factor, d1, d2);
+  input = THCTensor_(newContiguous)(state, input);
```
Comment (fmassa) [inline aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: same question as before: because we are using the `THCDeviceTensor`, and we index all the tensors using the `[]` operator, do we need to keep the restriction that the input is contiguous?

File: aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu [hunk 19]
```diff
@@ -5,151 +5,89 @@
...
+
+
 void THNN_(TemporalUpSamplingNearest_updateGradInput)(
            THCState *state,
-           THCTensor *input,
            THCTensor *gradOutput,
            THCTensor *gradInput,
-           int scale_factor)
+           int nbatch,
+	   int nchannels,
```
Comment (fmassa) [inline aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu]: nit: tabs here

File: aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu [hunk 20]
```diff
@@ -5,179 +5,107 @@
...
   }
 }
 
+
 void THNN_(VolumetricUpSamplingNearest_updateOutput)(
            THCState *state,
            THCTensor *input,
            THCTensor *output,
-           int scale_factor)
+	   int outputDepth,
```
Comment (fmassa) [inline aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu]: nit: tabs here

File: aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu [hunk 21]
```diff
@@ -5,179 +5,107 @@
...
+  THNN_(VolumetricUpSamplingNearest_shapeCheck)(state, input, NULL, nbatch, channels,
+		  inputDepth, inputHeight, inputWidth,
+		  outputDepth, outputHeight, outputWidth);
+  THAssert(inputDepth > 0 && inputHeight > 0 && inputWidth > 0 &&
+		  outputDepth > 0 && outputHeight > 0 && outputWidth > 0);
+
+  THCTensor_(resize5d)(state, output,
+                       THCTensor_(size)(state, input, 0),
+                       THCTensor_(size)(state, input, 1),
+		       outputDepth,
```
Comment (fmassa) [inline aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu]: nit: tabs in the next few lines

File: aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu [hunk 22]
```diff
@@ -5,179 +5,107 @@
...
+		       outputHeight,
+                       outputWidth);
+  THCTensor_(zero)(state, output);
 
   input = THCTensor_(newContiguous)(state, input);
```
Comment (fmassa) [inline aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu]: same question about contiguous, this could maybe be removed

File: aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu [hunk 23]
```diff
@@ -5,179 +5,107 @@
...
+
+
 void THNN_(VolumetricUpSamplingNearest_updateGradInput)(
            THCState *state,
-           THCTensor *input,
            THCTensor *gradOutput,
            THCTensor *gradInput,
-           int scale_factor)
+           int nbatch,
+	   int nchannels,
```
Comment (fmassa) [inline aten/src/THCUNN/generic/VolumetricUpSamplingNearest.cu]: tabs :-)

File: aten/src/THNN/generic/THNN.h [hunk 24]
```diff
@@ -750,13 +750,15 @@ TH_API void THNN_(TemporalUpSamplingNearest_updateOutput)(
...
-          int scale_factor);
+          int osizeW);
 TH_API void THNN_(TemporalUpSamplingNearest_updateGradInput)(
           THNNState *state,
-          THTensor *input,
           THTensor *gradOutput,
           THTensor *gradInput,
-          int scale_factor);
+          int isizeB,
+	  int isizeC,
```
Comment (fmassa) [inline aten/src/THNN/generic/THNN.h]: tabs :-)

File: aten/src/THNN/generic/TemporalUpSamplingNearest.c [hunk 25]
```diff
@@ -2,172 +2,133 @@
 #define TH_GENERIC_FILE "generic/TemporalUpSamplingNearest.c"
 #else
 
+#include "linear_upsampling.h"
+#include <stdio.h>
```
Comment (fmassa) [inline aten/src/THNN/generic/TemporalUpSamplingNearest.c]: do we need this include?

File: aten/src/THNN/generic/VolumetricUpSamplingNearest.c [hunk 26]
```diff
@@ -2,225 +2,177 @@
 #define TH_GENERIC_FILE "generic/VolumetricUpSamplingNearest.c"
 #else
 
+#include "linear_upsampling.h"
+#include <stdio.h>
```
Comment (fmassa) [inline aten/src/THNN/generic/VolumetricUpSamplingNearest.c]: unnecessary include?

File: test/test_nn.py [hunk 27]
```diff
@@ -4720,13 +4720,16 @@ def test_upsamplingLinear1d(self):
...
-            m = nn.Upsample(size=4, **kwargs)
-            in_t = torch.ones(1, 1, 2)
-            out_t = m(Variable(in_t))
-            self.assertEqual(torch.ones(1, 1, 4), out_t.data)
+            # test float scale factor up & downsampling
+            for scale_factor in [0.5, 1.5, 2]:
+                m = nn.Upsample(scale_factor=scale_factor, **kwargs)
+                in_t = torch.ones(1, 1, 2)
+                out_size = int(math.floor(in_t.shape[-1] * scale_factor))
+                out_t = m(Variable(in_t))
```
Comment (fmassa) [inline test/test_nn.py]: we can probably remove the `Variable` in the tests, there are a few more in this file that can be seen from the diff

File: torch/nn/modules/upsampling.py [hunk 28]
```diff
@@ -48,9 +48,12 @@ class Upsample(Module):
         0.3.1. Since then, the default behavior is ``align_corners = False``.
         See below for concrete examples on how this affects the outputs.
 
+    .. warning::
+        This class is deprecated in favor of :func:`~nn.functional.interpolate`.
```
Comment (fmassa) [inline torch/nn/modules/upsampling.py]: I have mixed feelings about deprecating `nn.Upsample` without giving a module alternative to the users.
In simple cases(like interpolating by a fixed scale factor), it might still be handy to provide a module.
Let's see what the others think.

**Review by fmassa — COMMENTED** [review]
I think this looks good to merge, thanks Ailing!

I have a bunch of minor nits that are not merge blockers, so feel free to postpone addressing them in a follow-up PR.

**Comment by fmassa** [comment]
I have the impression that the ROCM build failure is unrelated https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang3.8-rocmnightly-ubuntu16.04-build/1826//console , but might be better for others to have a look as well

**Comment by fmassa** [comment]
@pytorchbot retest this please

(let's see if it works with me :-) )

**Review by fmassa — APPROVED** [review]

**Review by soumith — APPROVED** [review]

**Comment by ailzhang** [comment]
This is merged by facebook-github-bot but somehow didn't close automatically. Closing
